python/revng/model/base.py

```diff
@@ -0,0 +1,170 @@
+from dataclasses import dataclass, fields
+from enum import Enum
+from typing import Generic, Type, TypeVar, get_args, get_origin, get_type_hints
+
+import yaml
+
+
+no_default = object()
+
+
+@dataclass
+class StructBase:
+    @classmethod
+    def from_dict(cls, **kwargs):
+        constructor_kwargs = {}
+
+        for field_name, field_value in kwargs.items():
+            field_spec = cls.__dataclass_fields__.get(field_name)
+            if field_spec is None:
+                raise ValueError(f"Field {field_name} is not allowed for type {cls.__name__}")
+
+            try:
+                # isinstance raises if field_spec.type is an instance of `typing._SpecialForm`
+                # such as `Optional[something]`
+                if isinstance(field_value, field_spec.type):
+                    constructor_kwargs[field_name] = field_value
+                    continue
+            except TypeError:
+                pass
```

Comment:
You sure `try` is the only solution here?

Reply:
Nope, I found a way more elegant solution